Tests with autotune.py fail on m1

A selection of tests are failing when ran on a parallel ubuntu desktop running on mac m1:

tests/test_autotuner.py .................F.                                                                                                            [  1%]
...
tests/test_buffering.py ..............x.........                                                                                                       [  3%]
...
tests/test_data.py .............................ss...                                                                                                  [ 11%]
...
tests/test_dse.py ......x............................................................................................................................. [ 47%]
.........FF............F............................................                                                                                   [ 51%]


Steps to reproduce problem:
1. run Linux ubuntu-linux-22-04-desktop 5.15.0-41-generic on parallel desktop
2. Install gcc python according to the version outlined below
3. python3 -m -venv  ~/env/devito_env
4. cd work_directory
5. source ~/env/devito_env/bin/activate
6. git clone https://github.com/devitocodes/devito.git
7. cd devito
8. pip install -e .
9. pytest -m "not parallel" -k "not adjoint" tests/

Host Hardware Description:
2021 14 inch Mac book Pro Apple M1 Max
Host OS Description:
Ventura 13.0

Virtual Environment Set-up:
Linux ubuntu-linux-22-04-desktop 5.15.0-41-generic #44-Ubuntu SMP Thu Jun 23 11:20:13 UTC 2022 aarch64 aarch64 aarch64 GNU/Linux
gcc (Ubuntu 11.2.0-19ubuntu1) 11.2.0
Python 3.10.6

Application used for virtualisation:
Parallels Desktop

Full program output is shown below
```
(devito_env) parallels@ubuntu-linux-22-04-desktop:~/Documents/Yr4/FYP/devito$ pytest -m "not parallel" -k "not adjoint" tests/ --maxfail 100
==================================================================== test session starts =====================================================================
platform linux -- Python 3.10.6, pytest-7.2.0, pluggy-1.0.0
rootdir: /home/parallels/Documents/Yr4/FYP/devito, configfile: setup.cfg
plugins: cov-4.0.0, nbval-0.9.6
collected 1797 items / 173 deselected / 4 skipped / 1624 selected                                                                                            

tests/test_autotuner.py .................F.                                                                                                            [  1%]
tests/test_benchmark.py .......                                                                                                                        [  1%]
tests/test_buffering.py ..............x.........                                                                                                       [  3%]
tests/test_builtins.py ...................................                                                                                             [  5%]
tests/test_caching.py ........................................................                                                                         [  8%]
tests/test_checkpointing.py .......                                                                                                                    [  9%]
tests/test_cinterface.py .                                                                                                                             [  9%]
tests/test_constant.py .                                                                                                                               [  9%]
tests/test_data.py .............................ss...                                                                                                  [ 11%]
tests/test_derivatives.py ............................................................................................................................ [ 18%]
............................................................................                                                                           [ 23%]
tests/test_differentiable.py ...                                                                                                                       [ 23%]
tests/test_dimension.py ...........................................................................                                                    [ 28%]
tests/test_dle.py .................................................................................................................................... [ 36%]
............................                                                                                                                           [ 38%]
tests/test_docstrings.py ....................                                                                                                          [ 39%]
tests/test_dse.py ......x............................................................................................................................. [ 47%]
.........FF............F............................................                                                                                   [ 51%]
tests/test_environment.py ..                                                                                                                           [ 51%]
tests/test_fission.py .....                                                                                                                            [ 52%]
tests/test_gradient.py ..................................                                                                                              [ 54%]
tests/test_iet.py ...........                                                                                                                          [ 55%]
tests/test_interpolation.py ............................                                                                                               [ 56%]
tests/test_ir.py .................................................................                                                                     [ 60%]
tests/test_linearize.py ..............                                                                                                                 [ 61%]
tests/test_lower_exprs.py .............                                                                                                                [ 62%]
tests/test_msparse.py ..........                                                                                                                       [ 63%]
tests/test_operator.py ............................................................................................................................... [ 70%]
.............                                                                                                                                          [ 71%]
tests/test_pickle.py .............................                                                                                                     [ 73%]
tests/test_resample.py .                                                                                                                               [ 73%]
tests/test_roundoff.py ................................................................                                                                [ 77%]
tests/test_save.py ..                                                                                                                                  [ 77%]
tests/test_skewing.py ...........                                                                                                                      [ 78%]
tests/test_staggered_utils.py ...................                                                                                                      [ 79%]
tests/test_subdomains.py .................                                                                                                             [ 80%]
tests/test_symbolic_coefficients.py ................................................................................................................   [ 87%]
tests/test_symbolics.py ...................................................................                                                            [ 91%]
tests/test_tensors.py ...............................................................................................                                  [ 97%]
tests/test_threading.py .                                                                                                                              [ 97%]
tests/test_timestepping.py .......                                                                                                                     [ 97%]
tests/test_tools.py ....................                                                                                                               [ 99%]
tests/test_tti.py ....                                                                                                                                 [ 99%]
tests/test_visitors.py ...........                                                                                                                     [100%]

========================================================================== FAILURES ==========================================================================
____________________________________________________________________ test_nested_nthreads ____________________________________________________________________

    @switchconfig(platform='knl7210')  # To trigger nested parallelism
    def test_nested_nthreads():
        grid = Grid(shape=(96, 96, 96))
        f = TimeFunction(name='f', grid=grid)
    
        op = Operator(Eq(f.forward, f.dx + 1.), openmp=True)
>       op.apply(time=10, autotune=True)

tests/test_autotuner.py:315: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
devito/operator/operator.py:806: in apply
    args = self.arguments(**kwargs)
devito/operator/operator.py:650: in arguments
    args = self._prepare_arguments(**kwargs)
devito/operator/operator.py:617: in _prepare_arguments
    args.update(self._autotune(args, autotune or configuration['autotuning']))
devito/core/operator.py:133: in _autotune
    args, summary = autotune(self, args, level, mode)
devito/core/autotuning.py:128: in autotune
    operator.cfunction(*list(at_args.values()))
devito/operator/operator.py:693: in cfunction
    self._jit_compile()
devito/operator/operator.py:678: in _jit_compile
    recompiled, src_file = self._compiler.jit_compile(self._soname,
devito/arch/compiler.py:322: in jit_compile
    _, _, _, recompiled = compile_from_string(self, target, code, src_file,
../../../../env/devito_env/lib/python3.10/site-packages/codepy/jit.py:433: in compile_from_string
    toolchain.build_extension(ext_file, source_paths, debug=debug)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = JITCompiler[GNUCompiler], ext_file = '/tmp/devito-jitcache-uid1000/f971691cedbd7832665a972e8a89a545e1166944.so'
source_files = ['/tmp/devito-jitcache-uid1000/f971691cedbd7832665a972e8a89a545e1166944.c'], debug = False

    def build_extension(self, ext_file, source_files, debug=False):
        cc_cmdline = (
                self._cmdline(source_files, False)
                + ["-o", ext_file]
                )
    
        from pytools.prefork import call
        if debug:
            print(" ".join(cc_cmdline))
    
        result = call(cc_cmdline)
    
        if result != 0:
            import sys
            print("FAILED compiler invocation:" + " ".join(cc_cmdline),
                  file=sys.stderr)
>           raise CompileError("module compilation failed")
E           codepy.CompileError: module compilation failed

../../../../env/devito_env/lib/python3.10/site-packages/codepy/toolchain.py:210: CompileError
-------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------
/tmp/devito-jitcache-uid1000/f971691cedbd7832665a972e8a89a545e1166944.c:10:10: fatal error: xmmintrin.h: No such file or directory
   10 | #include "xmmintrin.h"
      |          ^~~~~~~~~~~~~
compilation terminated.
FAILED compiler invocation:gcc -O3 -g -fPIC -Wall -std=c99 -march=native -Wno-unused-result -Wno-unused-variable -Wno-unused-but-set-variable -ffast-math -shared -fopenmp /tmp/devito-jitcache-uid1000/f971691cedbd7832665a972e8a89a545e1166944.c -lm -o /tmp/devito-jitcache-uid1000/f971691cedbd7832665a972e8a89a545e1166944.so
_______________________________________________ TestAliases.test_minimize_remainders_due_to_autopadding[False] _______________________________________________

self = <test_dse.TestAliases object at 0xffff99823bb0>, rotate = False

    @pytest.mark.parametrize('rotate', [False, True])
    @switchconfig(autopadding=True, platform='knl7210')  # Platform is to fix pad value
    def test_minimize_remainders_due_to_autopadding(self, rotate):
        """
        Check that the bounds of the Iteration computing an aliasing expression are
        relaxed (i.e., slightly larger) so that backend-compiler-generated remainder
        loops are avoided.
        """
        grid = Grid(shape=(3, 3, 3))
        x, y, z = grid.dimensions
        t = grid.stepping_dim
    
        u = TimeFunction(name='u', grid=grid, space_order=3)
        u1 = TimeFunction(name='u1', grid=grid, space_order=3)
    
        u.data_with_halo[:] = 0.5
        u1.data_with_halo[:] = 0.5
    
        # Leads to 3D aliases
        eqn = Eq(u.forward, _R(_R(u[t, x, y, z] + u[t, x+1, y+1, z+1])*3. +
                               _R(u[t, x+2, y+2, z+2] + u[t, x+3, y+3, z+3])*3. + 1.))
    
        op0 = Operator(eqn, opt=('noop', {'openmp': False}))
        op1 = Operator(eqn, opt=('advanced', {'openmp': False, 'cire-mingain': 0,
                                              'cire-rotate': rotate}))
    
        # Check code generation
        bns, pbs = assert_blocking(op1, {'x0_blk0'})
        xs, ys, zs = self.get_params(op1, 'x0_blk0_size', 'y0_blk0_size', 'z_size')
        arrays = [i for i in FindSymbols().visit(bns['x0_blk0']) if i.is_Array]
        assert len(arrays) == 1
        assert len(FindNodes(VExpanded).visit(pbs['x0_blk0'])) == 0
        assert arrays[0].padding == ((0, 0), (0, 0), (0, 30))
        self.check_array(arrays[0], ((1, 1), (1, 1), (1, 1)), (xs+2, ys+2, zs+32), rotate)
        # Check loop bounds
        trees = retrieve_iteration_tree(bns['x0_blk0'])
        assert len(trees) == 2
        expected_rounded = trees[0].inner
        assert expected_rounded.symbolic_max ==\
            z.symbolic_max + (z.symbolic_max - z.symbolic_min + 3) % 16 + 1
    
        # Check numerical output
        op0(time_M=1)
>       op1(time_M=1, u=u1)

tests/test_dse.py:1293: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
devito/operator/operator.py:740: in __call__
    return self.apply(**kwargs)
devito/operator/operator.py:811: in apply
    cfunction = self.cfunction
devito/operator/operator.py:693: in cfunction
    self._jit_compile()
devito/operator/operator.py:678: in _jit_compile
    recompiled, src_file = self._compiler.jit_compile(self._soname,
devito/arch/compiler.py:322: in jit_compile
    _, _, _, recompiled = compile_from_string(self, target, code, src_file,
../../../../env/devito_env/lib/python3.10/site-packages/codepy/jit.py:433: in compile_from_string
    toolchain.build_extension(ext_file, source_paths, debug=debug)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = JITCompiler[GNUCompiler], ext_file = '/tmp/devito-jitcache-uid1000/bc203367676f8724fd155796e8f3531871de5b50.so'
source_files = ['/tmp/devito-jitcache-uid1000/bc203367676f8724fd155796e8f3531871de5b50.c'], debug = False

    def build_extension(self, ext_file, source_files, debug=False):
        cc_cmdline = (
                self._cmdline(source_files, False)
                + ["-o", ext_file]
                )
    
        from pytools.prefork import call
        if debug:
            print(" ".join(cc_cmdline))
    
        result = call(cc_cmdline)
    
        if result != 0:
            import sys
            print("FAILED compiler invocation:" + " ".join(cc_cmdline),
                  file=sys.stderr)
>           raise CompileError("module compilation failed")
E           codepy.CompileError: module compilation failed

../../../../env/devito_env/lib/python3.10/site-packages/codepy/toolchain.py:210: CompileError
-------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------
Operator `Kernel` ran in 0.01 s
/tmp/devito-jitcache-uid1000/bc203367676f8724fd155796e8f3531871de5b50.c:10:10: fatal error: xmmintrin.h: No such file or directory
   10 | #include "xmmintrin.h"
      |          ^~~~~~~~~~~~~
compilation terminated.
FAILED compiler invocation:gcc -O3 -g -fPIC -Wall -std=c99 -march=native -Wno-unused-result -Wno-unused-variable -Wno-unused-but-set-variable -ffast-math -shared -fopenmp /tmp/devito-jitcache-uid1000/bc203367676f8724fd155796e8f3531871de5b50.c -lm -o /tmp/devito-jitcache-uid1000/bc203367676f8724fd155796e8f3531871de5b50.so
--------------------------------------------------------------------- Captured log call ----------------------------------------------------------------------
INFO     Devito:logger.py:114 Operator `Kernel` ran in 0.01 s
_______________________________________________ TestAliases.test_minimize_remainders_due_to_autopadding[True] ________________________________________________

self = <test_dse.TestAliases object at 0xffff99823c40>, rotate = True

    @pytest.mark.parametrize('rotate', [False, True])
    @switchconfig(autopadding=True, platform='knl7210')  # Platform is to fix pad value
    def test_minimize_remainders_due_to_autopadding(self, rotate):
        """
        Check that the bounds of the Iteration computing an aliasing expression are
        relaxed (i.e., slightly larger) so that backend-compiler-generated remainder
        loops are avoided.
        """
        grid = Grid(shape=(3, 3, 3))
        x, y, z = grid.dimensions
        t = grid.stepping_dim
    
        u = TimeFunction(name='u', grid=grid, space_order=3)
        u1 = TimeFunction(name='u1', grid=grid, space_order=3)
    
        u.data_with_halo[:] = 0.5
        u1.data_with_halo[:] = 0.5
    
        # Leads to 3D aliases
        eqn = Eq(u.forward, _R(_R(u[t, x, y, z] + u[t, x+1, y+1, z+1])*3. +
                               _R(u[t, x+2, y+2, z+2] + u[t, x+3, y+3, z+3])*3. + 1.))
    
        op0 = Operator(eqn, opt=('noop', {'openmp': False}))
        op1 = Operator(eqn, opt=('advanced', {'openmp': False, 'cire-mingain': 0,
                                              'cire-rotate': rotate}))
    
        # Check code generation
        bns, pbs = assert_blocking(op1, {'x0_blk0'})
        xs, ys, zs = self.get_params(op1, 'x0_blk0_size', 'y0_blk0_size', 'z_size')
        arrays = [i for i in FindSymbols().visit(bns['x0_blk0']) if i.is_Array]
        assert len(arrays) == 1
        assert len(FindNodes(VExpanded).visit(pbs['x0_blk0'])) == 0
        assert arrays[0].padding == ((0, 0), (0, 0), (0, 30))
        self.check_array(arrays[0], ((1, 1), (1, 1), (1, 1)), (xs+2, ys+2, zs+32), rotate)
        # Check loop bounds
        trees = retrieve_iteration_tree(bns['x0_blk0'])
        assert len(trees) == 2
        expected_rounded = trees[0].inner
        assert expected_rounded.symbolic_max ==\
            z.symbolic_max + (z.symbolic_max - z.symbolic_min + 3) % 16 + 1
    
        # Check numerical output
        op0(time_M=1)
>       op1(time_M=1, u=u1)

tests/test_dse.py:1293: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
devito/operator/operator.py:740: in __call__
    return self.apply(**kwargs)
devito/operator/operator.py:811: in apply
    cfunction = self.cfunction
devito/operator/operator.py:693: in cfunction
    self._jit_compile()
devito/operator/operator.py:678: in _jit_compile
    recompiled, src_file = self._compiler.jit_compile(self._soname,
devito/arch/compiler.py:322: in jit_compile
    _, _, _, recompiled = compile_from_string(self, target, code, src_file,
../../../../env/devito_env/lib/python3.10/site-packages/codepy/jit.py:433: in compile_from_string
    toolchain.build_extension(ext_file, source_paths, debug=debug)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = JITCompiler[GNUCompiler], ext_file = '/tmp/devito-jitcache-uid1000/10dcd0feb1ed8064e7db954d9f497944025b95ab.so'
source_files = ['/tmp/devito-jitcache-uid1000/10dcd0feb1ed8064e7db954d9f497944025b95ab.c'], debug = False

    def build_extension(self, ext_file, source_files, debug=False):
        cc_cmdline = (
                self._cmdline(source_files, False)
                + ["-o", ext_file]
                )
    
        from pytools.prefork import call
        if debug:
            print(" ".join(cc_cmdline))
    
        result = call(cc_cmdline)
    
        if result != 0:
            import sys
            print("FAILED compiler invocation:" + " ".join(cc_cmdline),
                  file=sys.stderr)
>           raise CompileError("module compilation failed")
E           codepy.CompileError: module compilation failed

../../../../env/devito_env/lib/python3.10/site-packages/codepy/toolchain.py:210: CompileError
-------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------
Operator `Kernel` ran in 0.01 s
/tmp/devito-jitcache-uid1000/10dcd0feb1ed8064e7db954d9f497944025b95ab.c:10:10: fatal error: xmmintrin.h: No such file or directory
   10 | #include "xmmintrin.h"
      |          ^~~~~~~~~~~~~
compilation terminated.
FAILED compiler invocation:gcc -O3 -g -fPIC -Wall -std=c99 -march=native -Wno-unused-result -Wno-unused-variable -Wno-unused-but-set-variable -ffast-math -shared -fopenmp /tmp/devito-jitcache-uid1000/10dcd0feb1ed8064e7db954d9f497944025b95ab.c -lm -o /tmp/devito-jitcache-uid1000/10dcd0feb1ed8064e7db954d9f497944025b95ab.so
--------------------------------------------------------------------- Captured log call ----------------------------------------------------------------------
INFO     Devito:logger.py:114 Operator `Kernel` ran in 0.01 s
__________________________________________________________ TestAliases.test_hoisting_symbolic_divs ___________________________________________________________

self = <test_dse.TestAliases object at 0xffff99822b30>

    def test_hoisting_symbolic_divs(self):
        grid = Grid(shape=(3, 3))
    
        f = Function(name='f', grid=grid, space_order=4)
        s0 = Scalar(name='s0')
        s1 = Scalar(name='s1')
    
        eq = Eq(f, f*(s0**-s1))
    
        op = Operator(eq)
    
        assert op._profiler._sections['section0'].sops == 1
>       assert op.body.body[1].body[0].body[0].expr.rhs == s0**-s1
E       IndexError: tuple index out of range

tests/test_dse.py:1629: IndexError
====================================================================== warnings summary ======================================================================
devito/finite_differences/differentiable.py:224: 5665 warnings
tests/test_autotuner.py: 12156 warnings
tests/test_buffering.py: 671 warnings
tests/test_builtins.py: 3674 warnings
tests/test_caching.py: 530 warnings
tests/test_checkpointing.py: 5454 warnings
tests/test_cinterface.py: 28 warnings
tests/test_constant.py: 62 warnings
tests/test_derivatives.py: 62800 warnings
tests/test_differentiable.py: 38 warnings
tests/test_dimension.py: 2352 warnings
tests/test_dle.py: 9582 warnings
tests/test_docstrings.py: 344 warnings
tests/test_dse.py: 204217 warnings
tests/test_fission.py: 439 warnings
tests/test_gradient.py: 155654 warnings
tests/test_iet.py: 46 warnings
tests/test_interpolation.py: 2411 warnings
tests/test_ir.py: 330 warnings
tests/test_linearize.py: 1571 warnings
tests/test_lower_exprs.py: 506 warnings
tests/test_msparse.py: 304 warnings
tests/test_operator.py: 5619 warnings
tests/test_pickle.py: 1015 warnings
tests/test_resample.py: 412 warnings
tests/test_roundoff.py: 4064 warnings
tests/test_save.py: 322 warnings
tests/test_skewing.py: 184 warnings
tests/test_staggered_utils.py: 827 warnings
tests/test_subdomains.py: 732 warnings
tests/test_symbolic_coefficients.py: 22145 warnings
tests/test_symbolics.py: 2046 warnings
tests/test_tensors.py: 28435 warnings
tests/test_threading.py: 16 warnings
tests/test_timestepping.py: 236 warnings
tests/test_tti.py: 28412 warnings
tests/test_visitors.py: 54 warnings
  /home/parallels/Documents/Yr4/FYP/devito/devito/finite_differences/differentiable.py:224: DeprecationWarning: NotImplemented should not be used in a boolean context
    return super(Differentiable, self).__eq__(other) and\

../../../../env/devito_env/lib/python3.10/site-packages/numpy/distutils/ccompiler.py:8
  /home/parallels/env/devito_env/lib/python3.10/site-packages/numpy/distutils/ccompiler.py:8: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
    from distutils import ccompiler

../../../../env/devito_env/lib/python3.10/site-packages/numpy/distutils/ccompiler.py:17
  /home/parallels/env/devito_env/lib/python3.10/site-packages/numpy/distutils/ccompiler.py:17: DeprecationWarning: The distutils.sysconfig module is deprecated, use sysconfig instead
    from distutils.sysconfig import customize_compiler

../../../../env/devito_env/lib/python3.10/site-packages/numpy/ctypeslib.py:137
  /home/parallels/env/devito_env/lib/python3.10/site-packages/numpy/ctypeslib.py:137: DeprecationWarning: 
  
    `numpy.distutils` is deprecated since NumPy 1.23.0, as a result
    of the deprecation of `distutils` itself. It will be removed for
    Python >= 3.12. For older Python versions it will remain present.
    It is recommended to use `setuptools < 60.0` for those Python versions.
    For more details, see:
      https://numpy.org/devdocs/reference/distutils_status_migration.html 
  
  
    from numpy.distutils.misc_util import get_shared_lib_extension

tests/test_autotuner.py::test_tti_aggressive[advanced]
tests/test_autotuner.py::test_tti_aggressive[opt1]
  /home/parallels/env/devito_env/lib/python3.10/site-packages/numpy/lib/arraysetops.py:272: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
    ar = np.asanyarray(ar)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
==================================================================== slowest 20 durations ====================================================================
9.60s call     tests/test_dse.py::TestTTI::test_opcounts[16-272]
9.30s call     tests/test_dse.py::TestTTI::test_fullopt
8.64s call     tests/test_dse.py::TestAliases::test_tti_J_akin_complete[so_ops1]
8.35s call     tests/test_threading.py::test_concurrent_executing_operators
5.11s call     tests/test_dse.py::TestTTIv2::test_opcounts[12-392]
4.72s call     tests/test_gradient.py::TestGradient::test_gradientFWI[float64-4-centered-shape2-True-tti_setup-2]
4.44s call     tests/test_gradient.py::TestGradient::test_gradient_checkpointing[OT2-shape0-spacing0-acoustic_setup-2-16-opt1-float64]
4.39s call     tests/test_gradient.py::TestGradient::test_gradientFWI[float32-4-centered-shape2-True-tti_setup-2]
4.31s call     tests/test_dse.py::TestIsoAcoustic::test_fullopt
4.29s call     tests/test_gradient.py::TestGradient::test_gradientFWI[float64-4-sls-shape4-True-viscoacoustic_setup-2]
4.27s call     tests/test_dse.py::TestTTI::test_opcounts[8-154]
4.15s call     tests/test_tti.py::test_tti[shape3-8-False]
4.04s call     tests/test_gradient.py::TestGradient::test_gradientFWI[float32-4-sls-shape4-True-viscoacoustic_setup-2]
3.85s call     tests/test_gradient.py::TestGradient::test_gradient_checkpointing[OT2-shape0-spacing0-acoustic_setup-2-16-opt0-float32]
3.76s call     tests/test_gradient.py::TestGradient::test_gradientFWI[float64-4-centered-shape3-False-tti_setup-2]
3.72s call     tests/test_benchmark.py::test_bench[run-viscoelastic-forward]
3.65s call     tests/test_tti.py::test_tti[shape2-4-True]
3.52s call     tests/test_gradient.py::TestGradient::test_gradientFWI[float64-4-sls-shape6-True-viscoacoustic_setup-1]
3.52s call     tests/test_gradient.py::TestGradient::test_gradientFWI[float64-4-sls-shape5-False-viscoacoustic_setup-2]
3.50s call     tests/test_gradient.py::TestGradient::test_gradientFWI[float32-4-centered-shape3-False-tti_setup-2]
================================================================== short test summary info ===================================================================
FAILED tests/test_autotuner.py::test_nested_nthreads - codepy.CompileError: module compilation failed
FAILED tests/test_dse.py::TestAliases::test_minimize_remainders_due_to_autopadding[False] - codepy.CompileError: module compilation failed
FAILED tests/test_dse.py::TestAliases::test_minimize_remainders_due_to_autopadding[True] - codepy.CompileError: module compilation failed
FAILED tests/test_dse.py::TestAliases::test_hoisting_symbolic_divs - IndexError: tuple index out of range
============================= 4 failed, 1616 passed, 6 skipped, 173 deselected, 2 xfailed, 563358 warnings in 368.81s (0:06:08) ==============================

```

--- a/docs/source/download.rst
+++ b/docs/source/download.rst
@@ -13,7 +13,7 @@ Docker
 For detailed installation instructions and information on the Devito Docker image library please follow 
 the docker/README.md_
 
-.. _README.md: ../../docker/README.md
+.. _README.md: https://github.com/devitocodes/devito/tree/master/docker#readme
 
 
 Virtual environment

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -6,7 +6,7 @@ import sys
 
 from devito import Eq, configuration  # noqa
 from devito.finite_differences.differentiable import EvalDerivative
-from devito.arch import Cpu64, Device, sniff_mpi_distro
+from devito.arch import Cpu64, Device, sniff_mpi_distro, Arm
 from devito.arch.compiler import compiler_registry, IntelCompiler, NvidiaCompiler
 from devito.ir.iet import retrieve_iteration_tree, FindNodes, Iteration, ParallelBlock
 from devito.tools import as_tuple
@@ -23,7 +23,7 @@ def skipif(items, whole_module=False):
     # Sanity check
     accepted = set()
     accepted.update({'device', 'device-C', 'device-openmp', 'device-openacc',
-                     'device-aomp', 'cpu64-icc', 'cpu64-nvc'})
+                     'device-aomp', 'cpu64-icc', 'cpu64-nvc', 'cpu64-arm'})
     accepted.update({'nompi', 'nodevice'})
     unknown = sorted(set(items) - accepted)
     if unknown:
@@ -68,6 +68,10 @@ def skipif(items, whole_module=False):
            isinstance(configuration['compiler'], IntelCompiler) and \
            isinstance(configuration['platform'], Cpu64):
             skipit = "`icc+cpu64` won't work with this test"
+            break
+        # Skip if it won't run on Arm
+        if i == 'cpu64-arm' and isinstance(configuration['platform'], Arm):
+            skipit = "Arm doesn't support x86-specific instructions"
             break
 
     if skipit is False:

--- a/tests/test_autotuner.py
+++ b/tests/test_autotuner.py
@@ -306,6 +306,7 @@ def test_multiple_threads(opt_options):
     assert len(op._state['autotuning'][0]['tuned']) == 3
 
 
+@skipif('cpu64-arm')
 @switchconfig(platform='knl7210')  # To trigger nested parallelism
 def test_nested_nthreads():
     grid = Grid(shape=(96, 96, 96))

--- a/tests/test_dse.py
+++ b/tests/test_dse.py
@@ -1253,6 +1253,7 @@ class TestAliases(object):
         assert np.all(u.data == u1.data)
         assert np.all(v.data == v1.data)
 
+    @skipif('cpu64-arm')
     @pytest.mark.parametrize('rotate', [False, True])
     @switchconfig(autopadding=True, platform='knl7210')  # Platform is to fix pad value
     def test_minimize_remainders_due_to_autopadding(self, rotate):
@@ -1632,7 +1633,7 @@ class TestAliases(object):
         op = Operator(eq)
 
         assert op._profiler._sections['section0'].sops == 1
-        assert op.body.body[1].body[0].body[0].expr.rhs == s0**-s1
+        assert op.body.body[-1].body[0].body[0].expr.rhs == s0**-s1
 
     @pytest.mark.parametrize('rotate', [False, True])
     def test_drop_redundants_after_fusion(self, rotate):

